src/bsoncxx/private/libbson_error.hpp

```diff
@@ -0,0 +1,54 @@
+// Copyright 2015 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <bsoncxx/config/prelude.hpp>
+
+#include <limits>
+#include <system_error>
+
+namespace bsoncxx {
+BSONCXX_INLINE_NAMESPACE_BEGIN
+
+enum class libbson_error : std::int32_t {
+    k_json_error_read_corrupt_js = 1,
+    k_json_error_read_invalid_param,
+    k_json_error_read_cb_failure,
+    k_bson_error_reader_badfd,
+    k_unknown = std::numeric_limits<std::int32_t>::max(),
+};
+
+///
+/// Convert an error code and domain from libbson into a libbson_error.
+///
+/// @param code A libbson error code
+/// @param domain A libbson error domain
+///
+/// @return A libbson error
+///
+libbson_error convert_libbson_error_code(int code, int domain);
```

Comment:
If each domain had a its own enum and category, then you would only need

`std::error_code make_error_code(int code, int domain)`